File: ios/Classes/YoutubePlayerPlugin.swift
// ios/Classes/YouTubePlayerPlugin.swift
import Flutter
import UIKit
import YouTubeiOSPlayerHelper

public class YouTubePlayerPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let factory = YouTubePlayerFactory(messenger: registrar.messenger())
    registrar.register(factory, withId: "youtube_player_view")
    NSLog("[NativeYouTubePlayer] Plugin registered")
  }
}

class YouTubePlayerFactory: NSObject, FlutterPlatformViewFactory {
  private let messenger: FlutterBinaryMessenger

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    super.init()
  }

  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    return NativeYouTubePlayer(
      frame: frame,
      viewIdentifier: viewId,
      arguments: args,
      messenger: messenger
    )
  }
}

class NativeYouTubePlayer: NSObject, FlutterPlatformView, YTPlayerViewDelegate {
  private let playerView: YTPlayerView
  private let channel: FlutterMethodChannel
  private var currentVideoId: String?
  // Store API key if provided from Dart.
  private var apiKey: String?

  init(
    frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?,
    messenger: FlutterBinaryMessenger
  ) {
    // Initialize playerView with the given frame.
    playerView = YTPlayerView(frame: frame)
    channel = FlutterMethodChannel(
      name: "youtube_player_\(viewId)",
      binaryMessenger: messenger
    )
    super.init()
    
    playerView.delegate = self
    setupMethodChannel()
    
    // Retrieve API key and videoId from creation parameters.
    if let params = args as? [String: Any] {
      apiKey = params["apiKey"] as? String
      if let videoId = params["videoId"] as? String {
        currentVideoId = videoId
      }
    }
    
    // Ensure an API key is provided.
    guard let validApiKey = apiKey, !validApiKey.isEmpty else {
      NSLog("[NativeYouTubePlayer] API key is missing! Please provide a valid API key from Dart.")
      channel.invokeMethod("onError", arguments: "API key is missing")
      return
    }
    
    // Initialize playerView with the provided API key.
    // FUTURE ENHANCEMENT: Consider a secure way to store and retrieve the API key.
    playerView.initialize(validApiKey, playerVars: ["playsinline": 1]) { [weak self] success, error in
      if success {
        NSLog("[NativeYouTubePlayer] YouTubePlayerView initialized successfully with API key")
        // If a videoId was provided in the creation parameters, load it now.
        if let videoId = self?.currentVideoId, !videoId.isEmpty {
          self?.loadVideo(videoId: videoId)
        }
      } else {
        let errMsg = error?.localizedDescription ?? "Unknown error"
        NSLog("[NativeYouTubePlayer] Error initializing player view: \(errMsg)")
        self?.channel.invokeMethod("onError", arguments: errMsg)
      }
    }
  }

  private func setupMethodChannel() {
    channel.setMethodCallHandler { [weak self] call, result in
      guard let self = self else { return }
      
      switch call.method {
      case "loadVideo":
        if let videoId = call.arguments as? String, !videoId.isEmpty {
          self.loadVideo(videoId: videoId)
          result(nil)
        } else {
          result(FlutterError(code: "INVALID_VIDEO_ID", message: "Video ID is null or empty", details: nil))
        }
      case "play":
        self.playerView.playVideo()
        result(nil)
      case "pause":
        self.playerView.pauseVideo()
        result(nil)
      case "seekTo":
        if let seconds = call.arguments as? Float {
          self.playerView.seek(toSeconds: seconds, allowSeekAhead: true)
          result(nil)
        } else {
          result(FlutterError(code: "INVALID_POSITION", message: "Position is null", details: nil))
        }
      case "setPlaybackRate":
        if let rate = call.arguments as? Float {
          self.playerView.setPlaybackRate(rate)
          result(nil)
        } else {
          result(FlutterError(code: "INVALID_RATE", message: "Playback rate is null", details: nil))
        }
      case "setQuality":
        if let quality = call.arguments as? String, !quality.isEmpty {
          self.playerView.setPlaybackQuality(quality)
          result(nil)
        } else {
          result(FlutterError(code: "INVALID_QUALITY", message: "Quality is null or empty", details: nil))
        }
      default:
        result(FlutterMethodNotImplemented)
      }
    }
  }

  private func loadVideo(videoId: String) {
    currentVideoId = videoId
    NSLog("[NativeYouTubePlayer] Loading video: \(videoId)")
    playerView.load(withVideoId: videoId)
  }

  func view() -> UIView {
    return playerView
  }

  // MARK: - YTPlayerViewDelegate Methods

  func playerViewDidBecomeReady(_ playerView: YTPlayerView) {
    NSLog("[NativeYouTubePlayer] Player ready")
    channel.invokeMethod("onReady", arguments: nil)
  }

  func playerView(_ playerView: YTPlayerView, didChangeTo state: YTPlayerState) {
    let stateString: String
    switch state {
    case .unstarted: stateString = "unstarted"
    case .ended: stateString = "ended"
    case .playing: stateString = "playing"
    case .paused: stateString = "paused"
    case .buffering: stateString = "buffering"
    case .queued: stateString = "queued"
    default: stateString = "unknown"
    }
    NSLog("[NativeYouTubePlayer] State changed: \(stateString)")
    channel.invokeMethod("onStateChange", arguments: stateString)
  }
  
  func playerView(_ playerView: YTPlayerView, didPlayTime playTime: Float) {
    let millis = Int(playTime * 1000)
    channel.invokeMethod("onTimeChange", arguments: millis)
  }
  
  func playerView(_ playerView: YTPlayerView, receivedError error: YTPlayerError) {
    let errorString = error.rawValue
    NSLog("[NativeYouTubePlayer] Received error: \(errorString)")
    channel.invokeMethod("onError", arguments: errorString)
  }
  
  // MARK: - Additional Recommendations & Future Enhancements:
  // 1. Consider adding buffering updates: If YTPlayerView provides buffering progress,
  //    relay this information via an additional method call (e.g., "onBuffering").
  // 2. Expose more granular event streams using Flutter's EventChannel if required.
  // 3. Implement full-screen toggling on the native side and expose that via MethodChannel.
  // 4. Provide configuration for controls, e.g., automatically hide native controls and
  //    let Flutter render custom overlays.
  // 5. Ensure that API key and sensitive data are handled securely.

  func dispose() {
    do {
      playerView.stopVideo()
      NSLog("[NativeYouTubePlayer] Stopped video playback")
      // YouTubeiOSPlayerHelper does not provide an explicit release method,
      // but you may perform additional cleanup if needed.
    } catch {
      NSLog("[NativeYouTubePlayer] Error stopping video: \(error.localizedDescription)")
    }
  }
}

File: android/src/main/kotlin/com/abwab/native_youtube_player/youtube_player/YoutubePlayerPlugin.kt
package com.abwab.native_youtube_player.youtube_player

import android.content.Context
import android.util.Log
import com.google.android.youtube.player.*
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

private const val TAG = "NativeYouTubePlayer"

// Main plugin class registering the view factory.
class YouTubePlayerPlugin : FlutterPlugin {
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "youtube_player")
        binding.platformViewRegistry.registerViewFactory(
            "youtube_player_view",
            YouTubePlayerFactory(binding.binaryMessenger)
        )
        Log.i(TAG, "YouTubePlayerPlugin attached to engine")
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        Log.i(TAG, "YouTubePlayerPlugin detached from engine")
    }
}

// Factory that creates instances of the native YouTube player view.
class YouTubePlayerFactory(private val messenger: io.flutter.plugin.common.BinaryMessenger) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val params = args as? Map<String, Any>
        return NativeYouTubePlayer(context, viewId, params, messenger)
    }
}

// Our main PlatformView wrapping the YouTubePlayerView.
class NativeYouTubePlayer(
    context: Context,
    viewId: Int,
    private val args: Map<String, Any>?,
    messenger: io.flutter.plugin.common.BinaryMessenger
) : PlatformView, YouTubePlayer.OnInitializedListener {

    private val playerView: YouTubePlayerView = YouTubePlayerView(context)
    private val channel = MethodChannel(messenger, "youtube_player_$viewId")
    private var youTubePlayer: YouTubePlayer? = null
    private var currentVideoId: String? = null

    init {
        setupMethodChannel()
        // FUTURE ENHANCEMENT: Pass API key securely from Dart via creationParams instead of hardcoding.
        val apiKey = args?.get("apiKey") as? String
        if (apiKey.isNullOrEmpty()) {
            Log.e(TAG, "API key is missing! Please supply a valid API key from Dart.")
            channel.invokeMethod("onError", "API key is missing")
        } else {
            try {
                // Initialize with the provided API key.
                playerView.initialize(apiKey, this)
                Log.i(TAG, "YouTubePlayerView initialized with API key")
            } catch (e: Exception) {
                Log.e(TAG, "Error initializing YouTubePlayerView", e)
                channel.invokeMethod("onError", e.localizedMessage)
            }
        }
    }

    private fun setupMethodChannel() {
        channel.setMethodCallHandler { call, result ->
            try {
                when (call.method) {
                    "loadVideo" -> {
                        val videoId = call.argument<String>("videoId")
                        if (videoId.isNullOrEmpty()) {
                            result.error("INVALID_VIDEO_ID", "Video ID is null or empty", null)
                        } else {
                            loadVideo(videoId)
                            result.success(null)
                        }
                    }
                    "play" -> {
                        youTubePlayer?.play()
                        result.success(null)
                    }
                    "pause" -> {
                        youTubePlayer?.pause()
                        result.success(null)
                    }
                    "seekTo" -> {
                        val position = call.argument<Int>("position")
                        if (position != null) {
                            youTubePlayer?.seekToMillis(position)
                            result.success(null)
                        } else {
                            result.error("INVALID_POSITION", "Position is null", null)
                        }
                    }
                    "setPlaybackRate" -> {
                        val rate = call.argument<Float>("rate")
                        if (rate != null) {
                            // Validate rate is one of the supported rates.
                            val playbackRate = when (rate) {
                                0.25f -> YouTubePlayer.PlaybackRate.RATE_0_25
                                0.5f -> YouTubePlayer.PlaybackRate.RATE_0_5
                                1.0f -> YouTubePlayer.PlaybackRate.RATE_1
                                1.5f -> YouTubePlayer.PlaybackRate.RATE_1_5
                                2.0f -> YouTubePlayer.PlaybackRate.RATE_2
                                else -> YouTubePlayer.PlaybackRate.RATE_1
                            }
                            youTubePlayer?.setPlaybackRate(playbackRate)
                            result.success(null)
                        } else {
                            result.error("INVALID_RATE", "Playback rate is null", null)
                        }
                    }
                    "setQuality" -> {
                        val quality = call.argument<String>("quality")
                        if (quality.isNullOrEmpty()) {
                            result.error("INVALID_QUALITY", "Quality is null or empty", null)
                        } else {
                            try {
                                // Uppercase quality to match enum names.
                                val qualityEnum = YouTubePlayer.PlaybackQuality.valueOf(quality.uppercase())
                                youTubePlayer?.setPlaybackQuality(qualityEnum)
                                result.success(null)
                            } catch (e: Exception) {
                                Log.e(TAG, "Invalid quality: $quality", e)
                                result.error("INVALID_QUALITY", e.localizedMessage, null)
                            }
                        }
                    }
                    else -> result.notImplemented()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error handling method call: ${call.method}", e)
                result.error("METHOD_CALL_ERROR", e.localizedMessage, null)
            }
        }
    }

    private fun loadVideo(videoId: String) {
        currentVideoId = videoId
        if (youTubePlayer == null) {
            // If the player isn't initialized yet, log a warning.
            Log.w(TAG, "Player not yet initialized; video will load when ready.")
        } else {
            Log.i(TAG, "Loading video: $videoId")
            youTubePlayer?.loadVideo(videoId)
        }
    }

    override fun onInitializationSuccess(
        provider: YouTubePlayer.Provider,
        player: YouTubePlayer,
        wasRestored: Boolean
    ) {
        youTubePlayer = player
        setupPlayerListeners(player)

        // If not restored, load the video from the creation params.
        if (!wasRestored) {
            args?.get("videoId")?.let {
                if (it is String && it.isNotEmpty()) {
                    currentVideoId = it
                    loadVideo(it)
                }
            }
        }
        channel.invokeMethod("onReady", null)
        Log.i(TAG, "Player initialized successfully")
    }

    override fun onInitializationFailure(
        provider: YouTubePlayer.Provider,
        error: YouTubeInitializationResult
    ) {
        Log.e(TAG, "Player initialization failed: $error")
        channel.invokeMethod("onError", error.toString())
        // FUTURE ENHANCEMENT: Implement a retry mechanism if initialization fails.
    }

    private fun setupPlayerListeners(player: YouTubePlayer) {
        player.setPlayerStateChangeListener(object : YouTubePlayer.PlayerStateChangeListener {
            override fun onLoading() {
                channel.invokeMethod("onStateChange", "loading")
                // FUTURE ENHANCEMENT: Consider sending buffering percentage if available.
            }
            
            override fun onLoaded(videoId: String) {
                channel.invokeMethod("onStateChange", "loaded")
            }
            
            override fun onVideoStarted() {
                channel.invokeMethod("onStateChange", "playing")
            }
            
            override fun onVideoEnded() {
                channel.invokeMethod("onStateChange", "ended")
            }
            
            override fun onError(error: YouTubePlayer.ErrorReason) {
                Log.e(TAG, "Player state error: $error")
                channel.invokeMethod("onError", error.toString())
            }

            override fun onAdStarted() {
                channel.invokeMethod("onStateChange", "ad_started")
            }
        })

        player.setPlaybackEventListener(object : YouTubePlayer.PlaybackEventListener {
            override fun onPlaying() {
                channel.invokeMethod("onStateChange", "playing")
            }
            
            override fun onPaused() {
                channel.invokeMethod("onStateChange", "paused")
            }
            
            override fun onStopped() {
                channel.invokeMethod("onStateChange", "stopped")
            }
            
            override fun onBuffering(isBuffering: Boolean) {
                channel.invokeMethod("onStateChange", if (isBuffering) "buffering" else "playing")
            }
            
            override fun onSeekTo(newPositionMillis: Int) {
                channel.invokeMethod("onTimeChange", newPositionMillis)
            }
        })
    }

    override fun getView() = playerView

    override fun dispose() {
        try {
            youTubePlayer?.release()
            Log.i(TAG, "Player released successfully")
        } catch (e: Exception) {
            Log.e(TAG, "Error releasing player", e)
        }
    }
}
